### ISSUE TITLE
Add focused property to control

### ISSUE THREAD

**Issue by Shadowblitz16** [godotengine/godot-proposals#6298]
### Describe the project you are working on

A controller based game

### Describe the problem or limitation you are having in your project

I would like to be able to set whether a control is focused by default when its ready

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

It would help avoid needing to add a script to a control node just to set it to focused on ready for keyboard and gamepad input.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

It would just be a bool whether or not it's focused.
on ready the control would just try to be focused unless another control grabs it.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

it can be but it's really unnecessary it should already be a property of the control node 

### Is there a reason why this should be core and not an add-on in the asset library?

It's simple and already part of the control node

**Comment by dalexeev** [comment]
I think a better name for this property is `autofocus` (like in HTML, like `Timer`'s `autostart` and `AudioStreamPlayer`'s `autoplay`).

But this can be easily implemented with two lines of script:

```gdscript
func _ready():
    grab_focus()
```

**Comment by Flavelius** [comment]
It could be useful to have this be a priority value, rather than a toggle, so that spontanteous, but implicitly priority based input consumers, like popups can work collaboratively:
textfield automatically starts with focus
popup opens, automatically gains focus because of higher priority, textfield automatically loses focus
popup closes, loses focus
textfield automatically regains focus

**Comment by KoBeWi** [comment]
What happens if multiple controls have this enabled? Only one control can be focused.

> func _ready():
    grab_focus()

This won't work reliably.
```GDScript
func _ready():
    grab_focus.call_deferred()
```
is better (it's even documented).

**Comment by Shadowblitz16** [comment]
> What happens if multiple controls have this enabled? Only one control can be focused.
> 
> > func _ready():
> > grab_focus()
> 
> This won't work reliably.
> 
> ```gdscript
> func _ready():
>     grab_focus.call_deferred()
> ```
> 
> is better (it's even documented).

if more then one control has it enabled then the last one who's ready will grab the focus.

**Comment by coppolaemilio** [comment]
Ran into this recently. I have different UI elements in different scenes, and when you jump to the new scene you lose the focus. It is fine if you are using a mouse, but when trying to play with a controller then you have to manually make a script for each scene to grab the focus of the place where it should start. 

I think there should be a "autofocus" or "grab focus" property to enable from the inspector.

If multiple elements have them, just go to the latest that gets called.

**Comment by Calinou** [comment]
> I think there should be a "autofocus" or "grab focus" property to enable from the inspector.

Properties can't have the same name as functions, so I'd go for `autofocus` which is consistent with the HTML attribute name.

**Comment by YuriSizov** [comment]
> if more then one control has it enabled then the last one who's ready will grab the focus.

Same will be true if we had a property. It's the same issue we have with cameras and their `current` property. So suggestions above are a completely valid solution.

**Comment by reduz** [comment]
I think "autofocus" is to vague

Maybe "show_focused" makes more sense?

**Comment by coppolaemilio** [comment]
> Maybe "show_focused" makes more sense?

I'm biased to like `autofocus` because of HTML. `show_focused` seems to be an option to see or hide the outline when an element is focused.

**Comment by reduz** [comment]
focus_on_show maybe?

**Comment by coppolaemilio** [comment]
haha feel a bit bad saying no to every proposal, but focus_on_show seems like when you toggle its visibility it will focus automatically, which doesn't seem to be the desired functionality. I like how the function grab_focus is called, because it will do just that, make the focus go to this, so if we want to keep the words, we could use "grab_focus_on_ready" but it is a bit too long. 
Maybe we can set up a poll and ask for different names, but I feel like most people will have the autofocus bias as well and default to that.

**Comment by YuriSizov** [comment]
For what it's worth, `autofocus` seems sensible to me. It can also be `autofocus_on_<something>` if we really want the name to be more descriptive. But I think the overall implication of the name `autofocus` is clear, and it cannot be confused with "focus" being used as a verb.

**Comment by reduz** [comment]
Yeah I guess if html people are used to this, we can just use it.

**Comment by reduz** [comment]
Maybe the FocusMode enum could be extended with a new value instead of adding a new property:
```
		FOCUS_NONE,
		FOCUS_CLICK, // when you click
		FOCUS_ALL, // when you click or tab
                FOCUS_AUTO // automatic
```

**Comment by ettiSurreal** [comment]
`focus_on_ready`? Might aswell throw something in.

**Comment by coppolaemilio** [comment]
> Maybe the FocusMode enum could be extended with a new value instead of adding a new property:
> 
> ```
> 		FOCUS_NONE,
> 		FOCUS_CLICK, // when you click
> 		FOCUS_ALL, // when you click or tab
>                 FOCUS_AUTO // automatic
> ```

Extending that Enum is probably not the solution because then you'll have to create
`FOCUS_AUTO_CLICK`, or `FOCUS_AUTO_ALL`, otherwise you won't have control over what kind of focus this node will capture. 

> focus_on_ready? Might aswell throw something in.

I like `focus_on_ready` 👍
